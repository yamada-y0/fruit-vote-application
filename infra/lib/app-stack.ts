import * as cdk from "aws-cdk-lib";
import { StackProps } from "aws-cdk-lib";
import * as ecr from "aws-cdk-lib/aws-ecr";
import { Construct } from "constructs";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as ecs from "aws-cdk-lib/aws-ecs";
import * as ecs_patterns from "aws-cdk-lib/aws-ecs-patterns";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import { BillingMode } from "aws-cdk-lib/aws-dynamodb";

export interface AppStackProps extends StackProps {
    appName: string;
    vpc: ec2.IVpc;
}

export class AppStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: AppStackProps) {
        super(scope, id, props);
        const repository = new ecr.Repository(this, "EcrRepository", {
            repositoryName: `${props.appName}-app`,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            emptyOnDelete: true,
        });

        const cluster = new ecs.Cluster(this, "EcsCluster", {
            vpc: props.vpc,
            clusterName: `${props.appName}-cluster`,
        });
        const fargateService =
            new ecs_patterns.ApplicationLoadBalancedFargateService(
                this,
                "FargateService",
                {
                    cluster: cluster,
                    cpu: 256,
                    desiredCount: 1,
                    taskImageOptions: {
                        image: ecs.ContainerImage.fromEcrRepository(repository),
                        containerPort: 8080,
                        enableLogging: true,
                    },
                    memoryLimitMiB: 512,
                    publicLoadBalancer: false,
                },
            );

        const voteTable = new dynamodb.Table(this, "VoteTable", {
            partitionKey: {
                name: "userId",
                type: dynamodb.AttributeType.STRING,
            },
            sortKey: {
                name: "fruit",
                type: dynamodb.AttributeType.STRING,
            },
            tableName: "Vote",
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            billingMode: BillingMode.PAY_PER_REQUEST,
        });
        voteTable.grantReadWriteData(
            fargateService.service.taskDefinition.taskRole,
        );

        const fruitCountTable = new dynamodb.Table(this, "FruitCountTable", {
            partitionKey: {
                name: "fruit",
                type: dynamodb.AttributeType.STRING,
            },
            tableName: "FruitCount",
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            billingMode: BillingMode.PAY_PER_REQUEST,
        });
        fruitCountTable.grantReadWriteData(
            fargateService.service.taskDefinition.taskRole,
        );
    }
}
