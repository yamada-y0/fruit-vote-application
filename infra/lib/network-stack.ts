import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as ec2 from "aws-cdk-lib/aws-ec2";

export class NetworkStack extends cdk.Stack {
    public readonly vpc: ec2.IVpc;

    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        this.vpc = new ec2.Vpc(this, "VPC", {
            ipAddresses: ec2.IpAddresses.cidr("10.0.0.0/16"),
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: "Isolated",
                    subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
                },
            ],
            maxAzs: 2,
        });

        new ec2.InterfaceVpcEndpoint(this, "EcrApiEndpoint", {
            vpc: this.vpc,
            service: ec2.InterfaceVpcEndpointAwsService.ECR,
        });
        new ec2.InterfaceVpcEndpoint(this, "EcrDockerEndpoint", {
            vpc: this.vpc,
            service: ec2.InterfaceVpcEndpointAwsService.ECR_DOCKER,
        });
        new ec2.InterfaceVpcEndpoint(this, "CloudWatchLogsEndpoint", {
            vpc: this.vpc,
            service: ec2.InterfaceVpcEndpointAwsService.CLOUDWATCH_LOGS,
        });
        new ec2.GatewayVpcEndpoint(this, "S3Endpoint", {
            vpc: this.vpc,
            service: ec2.GatewayVpcEndpointAwsService.S3,
        });
        new ec2.GatewayVpcEndpoint(this, "DynamoDbEndpoint", {
            vpc: this.vpc,
            service: ec2.GatewayVpcEndpointAwsService.DYNAMODB,
        });
    }
}
