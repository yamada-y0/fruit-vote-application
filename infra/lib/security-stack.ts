import * as cdk from 'aws-cdk-lib';
import {Duration} from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import {Construct} from 'constructs';

export class SecurityStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        const provider = new iam.OpenIdConnectProvider(this, 'GitHubOIDC', {
            url: 'https://token.actions.githubusercontent.com',
            clientIds: ['sts.amazonaws.com'],
        })
        const role = new iam.Role(this, 'GitHubActionsRole', {
            roleName: 'GitHubActionsRole',
            maxSessionDuration: Duration.hours(2),
            assumedBy: new iam.WebIdentityPrincipal(provider.openIdConnectProviderArn, {
                StringEquals: {
                    'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
                },
                StringLike: {
                    'token.actions.githubusercontent.com:sub': `repo:yamada-y0/fruit-vote-benchmark:*`,
                }
            }),
        })
        role.addManagedPolicy(
            iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonEC2ContainerRegistryPowerUser'),
        )
    }
}
