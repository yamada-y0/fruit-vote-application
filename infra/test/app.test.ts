import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as App from "../lib/app-stack";
import * as Network from "../lib/network-stack";

test("Application Resources Created", () => {
    const app = new cdk.App();
    // WHEN
    const networkStack = new Network.NetworkStack(app, "NetworkTestStack");
    const appStack = new App.AppStack(app, "AppTestStack", {
        appName: "quarkus",
        vpc: networkStack.vpc,
    });

    // THEN
    const template = Template.fromStack(appStack);

    template.hasResourceProperties("AWS::ECR::Repository", {
        RepositoryName: "quarkus-app",
        EmptyOnDelete: true,
    });
    template.hasResourceProperties("AWS::ECS::Cluster", {
        ClusterName: "quarkus-cluster",
    });
    template.hasResourceProperties("AWS::ECS::Service", {
        DesiredCount: 1,
    });
    template.hasResourceProperties("AWS::ECS::TaskDefinition", {
        ContainerDefinitions: [
            {
                PortMappings: [
                    {
                        ContainerPort: 8080,
                    },
                ],
            },
        ],
    });
});
