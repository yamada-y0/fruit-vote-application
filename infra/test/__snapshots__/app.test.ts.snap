// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppStack Test Snapshot Test 1`] = `
{
  "Outputs": {
    "FargateServiceLoadBalancerDNS9433D5F6": {
      "Value": {
        "Fn::GetAtt": [
          "FargateServiceLBB353E155",
          "DNSName",
        ],
      },
    },
    "FargateServiceServiceURL47701F45": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "http://",
            {
              "Fn::GetAtt": [
                "FargateServiceLBB353E155",
                "DNSName",
              ],
            },
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "EcrRepository4D7B3EE1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "EmptyOnDelete": true,
        "RepositoryName": "quarkus-app",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "EcsCluster97242B84": {
      "Properties": {
        "ClusterName": "quarkus-cluster",
      },
      "Type": "AWS::ECS::Cluster",
    },
    "FargateServiceECC8084D": {
      "DependsOn": [
        "FargateServiceLBPublicListenerECSGroupBE57E081",
        "FargateServiceLBPublicListener4B4929CA",
        "FargateServiceTaskDefTaskRoleDefaultPolicy63F83D6F",
        "FargateServiceTaskDefTaskRole8CDCF85E",
      ],
      "Properties": {
        "Cluster": {
          "Ref": "EcsCluster97242B84",
        },
        "DeploymentConfiguration": {
          "Alarms": {
            "AlarmNames": [],
            "Enable": false,
            "Rollback": false,
          },
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 50,
        },
        "DesiredCount": 1,
        "EnableECSManagedTags": false,
        "HealthCheckGracePeriodSeconds": 60,
        "LaunchType": "FARGATE",
        "LoadBalancers": [
          {
            "ContainerName": "web",
            "ContainerPort": 8080,
            "TargetGroupArn": {
              "Ref": "FargateServiceLBPublicListenerECSGroupBE57E081",
            },
          },
        ],
        "NetworkConfiguration": {
          "AwsvpcConfiguration": {
            "AssignPublicIp": "DISABLED",
            "SecurityGroups": [
              {
                "Fn::GetAtt": [
                  "FargateServiceSecurityGroup262B61DD",
                  "GroupId",
                ],
              },
            ],
            "Subnets": [
              {
                "Fn::ImportValue": "NetworkTestStack:ExportsOutputRefVPCIsolatedSubnet1SubnetEBD00FC6298E81EF",
              },
              {
                "Fn::ImportValue": "NetworkTestStack:ExportsOutputRefVPCIsolatedSubnet2Subnet4B1C8CAAD8B83B81",
              },
            ],
          },
        },
        "TaskDefinition": {
          "Ref": "FargateServiceTaskDef940E3A80",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "FargateServiceLBB353E155": {
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internal",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "FargateServiceLBSecurityGroup5F444C78",
              "GroupId",
            ],
          },
        ],
        "Subnets": [
          {
            "Fn::ImportValue": "NetworkTestStack:ExportsOutputRefVPCIsolatedSubnet1SubnetEBD00FC6298E81EF",
          },
          {
            "Fn::ImportValue": "NetworkTestStack:ExportsOutputRefVPCIsolatedSubnet2Subnet4B1C8CAAD8B83B81",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "FargateServiceLBPublicListener4B4929CA": {
      "Properties": {
        "DefaultActions": [
          {
            "TargetGroupArn": {
              "Ref": "FargateServiceLBPublicListenerECSGroupBE57E081",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "FargateServiceLBB353E155",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "FargateServiceLBPublicListenerECSGroupBE57E081": {
      "Properties": {
        "Port": 80,
        "Protocol": "HTTP",
        "TargetGroupAttributes": [
          {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "ip",
        "VpcId": {
          "Fn::ImportValue": "NetworkTestStack:ExportsOutputRefVPCB9E5F0B4BD23A326",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "FargateServiceLBSecurityGroup5F444C78": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB AppTestStackFargateServiceLB3EEA14BA",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "NetworkTestStack:ExportsOutputRefVPCB9E5F0B4BD23A326",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "FargateServiceLBSecurityGrouptoAppTestStackFargateServiceSecurityGroup883A707080801FA66932": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "FargateServiceSecurityGroup262B61DD",
            "GroupId",
          ],
        },
        "FromPort": 8080,
        "GroupId": {
          "Fn::GetAtt": [
            "FargateServiceLBSecurityGroup5F444C78",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 8080,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "FargateServiceSecurityGroup262B61DD": {
      "DependsOn": [
        "FargateServiceTaskDefTaskRoleDefaultPolicy63F83D6F",
        "FargateServiceTaskDefTaskRole8CDCF85E",
      ],
      "Properties": {
        "GroupDescription": "AppTestStack/FargateService/Service/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "NetworkTestStack:ExportsOutputRefVPCB9E5F0B4BD23A326",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "FargateServiceSecurityGroupfromAppTestStackFargateServiceLBSecurityGroupDD23365480801DD6E28B": {
      "DependsOn": [
        "FargateServiceTaskDefTaskRoleDefaultPolicy63F83D6F",
        "FargateServiceTaskDefTaskRole8CDCF85E",
      ],
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 8080,
        "GroupId": {
          "Fn::GetAtt": [
            "FargateServiceSecurityGroup262B61DD",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "FargateServiceLBSecurityGroup5F444C78",
            "GroupId",
          ],
        },
        "ToPort": 8080,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "FargateServiceTaskDef940E3A80": {
      "Properties": {
        "ContainerDefinitions": [
          {
            "Essential": true,
            "Image": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Select": [
                      4,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "EcrRepository4D7B3EE1",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".dkr.ecr.",
                  {
                    "Fn::Select": [
                      3,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "EcrRepository4D7B3EE1",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".",
                  {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  {
                    "Ref": "EcrRepository4D7B3EE1",
                  },
                  ":latest",
                ],
              ],
            },
            "LogConfiguration": {
              "LogDriver": "awslogs",
              "Options": {
                "awslogs-group": {
                  "Ref": "FargateServiceTaskDefwebLogGroup71FAF541",
                },
                "awslogs-region": {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "FargateService",
              },
            },
            "Name": "web",
            "PortMappings": [
              {
                "ContainerPort": 8080,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "Cpu": "256",
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "FargateServiceTaskDefExecutionRole9194820E",
            "Arn",
          ],
        },
        "Family": "AppTestStackFargateServiceTaskDefE42ECBF3",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE",
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "FargateServiceTaskDefTaskRole8CDCF85E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "FargateServiceTaskDefExecutionRole9194820E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "FargateServiceTaskDefExecutionRoleDefaultPolicy827E7CA2": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "EcrRepository4D7B3EE1",
                  "Arn",
                ],
              },
            },
            {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "FargateServiceTaskDefwebLogGroup71FAF541",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FargateServiceTaskDefExecutionRoleDefaultPolicy827E7CA2",
        "Roles": [
          {
            "Ref": "FargateServiceTaskDefExecutionRole9194820E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FargateServiceTaskDefTaskRole8CDCF85E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "FargateServiceTaskDefTaskRoleDefaultPolicy63F83D6F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "VoteTableC0BC27A7",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "FruitCountTable071E859E",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FargateServiceTaskDefTaskRoleDefaultPolicy63F83D6F",
        "Roles": [
          {
            "Ref": "FargateServiceTaskDefTaskRole8CDCF85E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FargateServiceTaskDefwebLogGroup71FAF541": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "FruitCountTable071E859E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "fruit",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "fruit",
            "KeyType": "HASH",
          },
        ],
        "TableName": "FruitCount",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "VoteTableC0BC27A7": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "userId",
            "AttributeType": "S",
          },
          {
            "AttributeName": "fruit",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "userId",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "fruit",
            "KeyType": "RANGE",
          },
        ],
        "TableName": "Vote",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
